samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+          + "right side of join should be a 'table'. " + dumpRelPlanForNode(join));
+    }
+
+    if (joinRelType.compareTo(JoinRelType.RIGHT) == 0 && isTablePosOnRight && !isTablePosOnLeft) {
+      throw new SamzaException("Invalid query for outer right join. Left side of the join should be a 'table' and "
+          + "right side of join should be a 'stream'. " + dumpRelPlanForNode(join));
+    }
+
+    validateJoinCondition(join.getCondition());
+  }
+
+  private void validateJoinCondition(RexNode operand) {
+    if (!(operand instanceof RexCall)) {
+      throw new SamzaException("SQL Query is not supported. Join condition operand " + operand +
+          " is of type " + operand.getClass());
+    }
+
+    RexCall condition = (RexCall) operand;
+
+    if (condition.isAlwaysTrue()) {
+      throw new SamzaException("Query results in a cross join, which is not supported. Please optimize the query."
+          + " It is expected that the joins should include JOIN ON operator in the sql query.");
+    }
+
+    if (condition.getKind() != SqlKind.EQUALS && condition.getKind() != SqlKind.AND) {
+      throw new SamzaException("Only equi-joins and AND operator is supported in join condition.");
+    }
+  }
+
+  // Get the stream and table indices corresponding to the fields given in the join condition by parsing through
+  // the condition.
+  private void getStreamAndTableKeyIds(List<RexNode> operands, final LogicalJoin join, boolean isTablePosOnRight,
+      List<Integer> streamIds, List<Integer> tableIds) {
+
+    // All non-leaf operands in the join condition should be expressions.
+    if (operands.get(0) instanceof RexCall) {
+      operands.forEach(operand -> {
+        validateJoinCondition(operand);
+        getStreamAndTableKeyIds(((RexCall) operand).getOperands(), join, isTablePosOnRight, streamIds, tableIds);
+      });
+      return;
+    }
+
+    // We are at the leaf of the join condition. Only binary operators are supported.
+    Validate.isTrue(operands.size() == 2);
+
+    // Only reference operands are supported in row expressions and not constants.
+    // a.key = b.key is supported with a.key and b.key being reference operands.
+    // a.key = "constant" is not yet supported.
+    if (!(operands.get(0) instanceof RexInputRef) || !(operands.get(1) instanceof RexInputRef)) {
+      throw new SamzaException("SQL query is not supported. Join condition " + join.getCondition() + " should have "
+          + "reference operands but the types are " + operands.get(0).getClass() + " and " + operands.get(1).getClass());
+    }
+
+    // Join condition is commutative, meaning, a.key = b.key is equivalent to b.key = a.key.
+    // Calcite assigns the indices to the fields based on the order a and b are specified in
+    // the sql 'from' clause. Let's put the operand with smaller index in leftRef and larger
+    // index in rightRef.
```

Comment:
How does this work when the query contains join condition like select * from a join b on a.key=b.key and b.field1=a.field1?

Reply:
The comment is an attempt to explain that :) Operand with smaller index are collected in leftRef while the other in rightRef. In the example that you gave, a.key and a.field1 are collected in leftRef while b.key and b.field1 are collected in rightRef. Let me see if I can make this clear in the comment.